resources/2024-07-secretstore-feature-spec.md

```diff
@@ -0,0 +1,286 @@
+    container: {
+      image: 'ghcr.io/radius-project/samples/demo:latest'
+      env:{
++        DB_CONNECTION: {
++          valueFrom: {
++            secretRef: {
++              source: authcreds.id
++              key: 'username'
++            }
++          }
++        }
+      }
+      ports: {
+        web: {
+          containerPort: 3000
+        }
+      }
+    }
+  }
+}
+```
+
+Step 3: Developer references the `Applications.Core/secretStores` resource in their `Applications.Extenders` or `Applications.Volumes` resource definition to securely manage secrets for use in their application.
+
+```diff
+resource twilio 'Applications.Core/extenders@2023-10-01-preview' = {
+  name: 'twilio'
+  properties: {
+    application: application
+    environment: environment
+    recipe: {
+      name: 'twilio'
+    }
+    secrets: {
++        password: {
++          valueFrom: {
++            secretRef: {
++              source: authcreds.id
++              key: 'password'
++            }
++          }
++        }
+    }
+  }
+}
+```
+
+```diff
+resource volume 'Applications.Core/volumes@2023-10-01-preview' = {
+  name: 'myvolume'
+  properties: {
+    application: app.id
+    kind: 'azure.com.keyvault'
+    resource: keyvault.id
+    secrets: {
+      mysecret: {
++        name: {
++          valueFrom: {
++            secretRef: {
++              source: azurekeyvaultsecrets.id
++              key: 'name'
++            }
++          }
++        }
++        version: {
++          valueFrom: {
++            secretRef: {
++              source: azurekeyvaultsecrets.id
++              key: 'version'
++            }
++          }
++        }
++        alias: {
++          valueFrom: {
++            secretRef: {
++              source: azurekeyvaultsecrets.id
++              key: 'alias'
++            }
++          }
++        }
++        encoding: {
++          valueFrom: {
++            secretRef: {
++              source: azurekeyvaultsecrets.id
++              key: 'encoding'
++            }
++          }
++        }
+      }
+    }
+  }
+}
+```
+
+Step 4: Developer references the `Applications.Core/secretStores` resource in their `Applications.Datastores/*` or `Applications.Messaging/*` resource definition to securely manage secrets for use in their application.
+
+```diff
+resource db 'Applications.Datastores/mongoDatabases@2023-10-01-preview' = {
+  name: 'db'
+  properties: {
+    environment: environment
+    application: app.id
+    resourceProvisioning: 'manual'
+    host: substring(cosmosAccount.properties.documentEndpoint, 0, lastIndexOf(cosmosAccount.properties.documentEndpoint, ':'))
+    port: int(split(substring(cosmosAccount.properties.documentEndpoint,lastIndexOf(cosmosAccount.properties.documentEndpoint, ':') + 1), '/')[0])
+    database: cosmosAccount::database.name
+    username: ''
+    resources: [
+      { id: cosmosAccount.id }
+    ]
+    secrets: {
++        connectionString: {
++          valueFrom: {
```

Comment:
We should revisit the overall plan for provisioning types we plan to support on portable resources before implementing this change. There are some changes with UDT that haven't been solidified yet.

Reply:
added this as a note to the feature spec